### ISSUE TITLE
systemd-sysusers and Fedora/RHEL

### ISSUE THREAD

**Issue by frankcrawford** [rfmoz/tuptime#86]
Have you given any thoughts to the changes to the RPM spec file to make use of systemd-sysusers, similar to what has been done for Debian?

It should be a pretty simple change, but I'm reluctant to push it straight into Fedora without some prior testing, so I won't use it for current version.

Also, now that RHEL7 has gone, RHEL8 and on should also support it.

**Comment by rfmoz** [comment]
I've checked how systemd-sysusers apply on RPM and [their doc](https://rpm-software-management.github.io/rpm/manual/users_and_groups.html) said that it is available in Rpm >= 4.19  so Fedora 39 and later, RHEL 8 I guess that uses 4.14, so RHEL 9 and later.

As the current tuptime rpm package works fine and it's compatible with a large number of releases, maybe introducing sysusers could narrow the scope. I haven't done anything because of that, feel free to apply it, you know better what could be the impact.

On the Debian side, the .deb package will be updated on the official repos in a few weeks. I'm expected to monitor its progress (I hope everything goes smoothly).

**Comment by frankcrawford** [comment]
What I will do is upgrade to the new version, using the current spec file, and play around with the new systemd-sysusers format in COPR after that.

It is interesting that RPM upgrade to internally support systemd-sysusers is only in Fedora, but doing an installation similar to your installation script without using the RPM macro should work on RHEL systems.

**Comment by rfmoz** [comment]
Ok, it seems a good approach. Please, let me know if you need any help.

**Comment by rfmoz** [comment]
Using the attached spec file ([tuptime.spec.txt](https://github.com/user-attachments/files/16767514/tuptime.spec.txt)), I've built and installed the package using sysusers in Fedora 40 without problem.

I followed the docu on [https://docs.fedoraproject.org/en-US/packaging-guidelines/UsersAndGroups/](https://docs.fedoraproject.org/en-US/packaging-guidelines/UsersAndGroups/) and [https://github.com/openela-main/munge/blob/el9/SPECS/munge.spec](https://github.com/openela-main/munge/blob/el9/SPECS/munge.spec) as example.

Please, could you take a look?

**Comment by frankcrawford** [comment]
On a quick look, it seems to be what I saw from some of the docs you reference, although I may need to split it into the case for EPEL vs everything else that handles the new macros.

But first, I'll test it out to make sure it works as expected, and then go to handling compatibility issues with older systems.

BTW, as EPEL7 is no longer available, I guess I really only need to worry about EPEL8 & 9, which means I may strip out some of the other tests too.

**Comment by frankcrawford** [comment]
Sorry I've left this for so long, but it is both because I've been busy and because Fedora devs have been discussing enforcing this creation method going forward anyway.

As such, how it is done now has simplified even more, and I'm now going to try and add it for use in F42 and maybe for the other current releases.

**Comment by rfmoz** [comment]
Great, don't worry.

The removal of:
```
%post
# Create and initialise the tuptime DB with consistent permissions, etc.
su -s /bin/sh _tuptime -c "(umask 0022 && /usr/bin/tuptime -q)"
%systemd_post tuptime.service
%systemd_post tuptime-sync.service
%systemd_post tuptime-sync.timer
```
it's related with the `%pre` section and the usage of sysusers.

As the new version lacks of useradd/groupadd and relies on sysusers, the user/group _tuptime is not available at the right time when the package is installed. If you keep that line to a testing .spec, it is executed just before user creation:

```
Running transaction
  Preparing        :                                                                                                                                                                      1/1 
  Running scriptlet: tuptime-5.2.4-2.fc40.noarch                                                                                                                                          1/1 
  Installing       : tuptime-5.2.4-2.fc40.noarch                                                                                                                                          1/1 
warning: user _tuptime does not exist - using root
warning: group _tuptime does not exist - using root

  Running scriptlet: tuptime-5.2.4-2.fc40.noarch                                                                                                                                          1/1 
su: user _tuptime does not exist or the user entry does not contain all the required fields

Creating group '_tuptime' with GID 993.
Creating user '_tuptime' (Tuptime execution user) with UID 993 and GID 993.


Installed:
  tuptime-5.2.4-2.fc40.noarch    
```

It doesn't matter if it is placed after the systemd calls::

```
%post
%systemd_post tuptime.service
%systemd_post tuptime-sync.service
%systemd_post tuptime-sync.timer
# Create and initialise the tuptime DB with consistent permissions, etc.
su -s /bin/sh _tuptime -c "(umask 0022 && /usr/bin/tuptime -q)"
```

```Running transaction
  Preparing        :                                                                                                                                                                      1/1 
  Running scriptlet: tuptime-5.2.4-2.fc40.noarch                                                                                                                                          1/1 
  Installing       : tuptime-5.2.4-2.fc40.noarch                                                                                                                                          1/1 
warning: user _tuptime does not exist - using root
warning: group _tuptime does not exist - using root

  Running scriptlet: tuptime-5.2.4-2.fc40.noarch                                                                                                                                          1/1 
su: user _tuptime does not exist or the user entry does not contain all the required fields
warning: %post(tuptime-5.2.4-2.fc40.noarch) scriptlet failed, exit status 1

Error in POSTIN scriptlet in rpm package tuptime
Creating group '_tuptime' with GID 993.
Creating user '_tuptime' (Tuptime execution user) with UID 993 and GID 993.


Installed:
  tuptime-5.2.4-2.fc40.noarch                                                                                                                                                                 

Complete!
```
I also trying with `%posttrans`, but the result is equal:
```su: user _tuptime does not exist or the user entry does not contain all the required fields
warning: %posttrans(tuptime-5.2.4-2.fc40.noarch) scriptlet failed, exit status 1

Error in POSTTRANS scriptlet in rpm package tuptime
Creating group '_tuptime' with GID 993.
Creating user '_tuptime' (Tuptime execution user) with UID 993 and GID 993.
```

Nevertheless, when the unit is started, `/var/lib/tuptime/tuptime.db` is created with _tuptime:_tuptime as expected. But... if root executes `tuptime` before first start, the db file is created belonging to root user.

**Comment by frankcrawford** [comment]
Yes, I thought that might be the case.  I've seen other comments that seemed to imply the user wasn't available during the post-install scripts, so there may be no way to fix it.

However, as you noted, it probably doesn't really affect anything in normal use.

**Comment by frankcrawford** [comment]
Looking at these comments, it may work properly in the near future:

```
> – are run after the RPM installation.

The new mechinism in rpm itself creates the accounts _before_ the rpm
is installed. (When the rpm is being built, if it has a sysusers file,
a virtual Provides is generated. When rpm sees an rpm with the Provides,
it will precreate the account.)

So yeah, having a package with a sysusers file and files owned by the
users or groups defined therein works fine. (This was already possible
before, but required careful creation of a %pre scriptlet. The new
mechanism is much nicer.)
```

**Comment by frankcrawford** [comment]
And delving even further, this will come in with Fedora 42, which is due to be released in a few months.

It is in Fedora rawhide now, which is what I've built for, and so going forward we can actually add that initialisation back in, if you want.

However, it does look like it won't be in RHEL10, as that is rpm 4.19, which is not fully functional with these features anyway.

**Comment by rfmoz** [comment]
Well, I think that the problem is related with a corner case behaviour and it's better to be aligned with the upcoming releases.

**Comment by frankcrawford** [comment]
Okay, so I'll set up the Fedora/EPEL spec file to use the old method for Fedora <= 41 and EPEL <= 10, and add back in the initialisation, and then do a PR to update your one in the distribution to match.

**Comment by frankcrawford** [comment]
I've had a look and it shouldn't be an issue.  I'll leave out the commented initialisation from the standard just to keep it cleaner, until you find a different way to do it, but I'll drop that extra set of examples, etc.

**Comment by rfmoz** [comment]
Thanks Frank, it should work as it is.
